### ISSUE TITLE
Support for multiple instead of just one `@children` slot

### ISSUE THREAD

**Issue by Enyium** [slint-ui/slint#6258]
Vue.js has a [slot concept](https://vuejs.org/guide/components/slots.html#named-slots).

The following code imagines how this concept could manifest in Slint:

```slint
component Page {
    VerticalLayout {
        Rectangle {
            background: gray;
            @children["header"] // Or `@children[0]`.
        }

        Rectangle {
            border-width: 1px;
            border-color: black;
            VerticalLayout {
                @children["body"]
                // Or `@children[1]`.
                // Or just `@children`, if it'd be named "default" (`slot`'s/`name`'s default value).
            }
        }

        if @has-slot("footer") : {
            SeparatorLine {}
            @children["footer"] // Or `@children[2]`.
        } else {
            Text { text: "Fallback Footer"; }
        }
    }
}

export component Demo {
    Page {
        Text {
            slot: "header"; // This property can be used on any element.
            text: "The Header";
        }
        Slot {
            name: "body";
            Text { text: "Some Content"; }
            Text { text: "More Content"; }
        }
        Slot {
            name: "footer";
            Text { text: "The Footer"; }
        }
    }
}
```

This feature would allow for better abstractions.

[This SlintPad demo](https://slintpad.com/?snippet=import+%7B+StyleMetrics%2C+Button%2C+VerticalBox%2C+HorizontalBox+%7D+from+%22std-widgets.slint%22%3B%0A%0Acomponent+Triad+%7B%0A++++preferred-width%3A+%28%0A++++++++2+*+max%28%0A++++++++++++left.preferred-width%2C%0A++++++++++++right.preferred-width%2C%0A++++++++%29+%2B%0A++++++++center.preferred-width+%2B%0A++++++++2+*+StyleMetrics.layout-spacing%0A++++%29%3B%0A++++preferred-height%3A+max%28%0A++++++++left.preferred-height%2C%0A++++++++center.preferred-height%2C%0A++++++++right.preferred-height%2C%0A++++%29%3B%0A%0A++++min-width%3A+%28%0A++++++++2+*+max%28%0A++++++++++++left-inner.width%2C%0A++++++++++++right-inner.width%2C%0A++++++++%29+%2B%0A++++++++center-inner.width+%2B%0A++++++++2+*+StyleMetrics.layout-spacing%0A++++%29%3B%0A++++min-height%3A+max%28%0A++++++++left-inner.height%2C%0A++++++++center-inner.height%2C%0A++++++++right-inner.height%2C%0A++++%29%3B%0A%0A++++left+%3A%3D+Rectangle+%7B%0A++++++++x%3A+0%3B%0A++++++++y%3A+0%3B%0A++++++++width%3A+%28root.width+-+center.width%29+%2F+2+-+StyleMetrics.layout-spacing%3B%0A++++++++background%3A+red%3B%0A%0A++++++++left-inner+%3A%3D+Rectangle+%7B%0A++++++++++++width%3A+20px%3B%0A++++++++++++height%3A+20px%3B%0A++++++++++++background%3A+%23fff8%3B%0A++++++++%7D%0A++++%7D%0A%0A++++center+%3A%3D+Rectangle+%7B%0A++++++++x%3A+%28root.width+-+self.width%29+%2F+2%3B%0A++++++++y%3A+0%3B%0A++++++++width%3A+center-inner.width%3B%0A++++++++background%3A+lime%3B%0A%0A++++++++center-inner+%3A%3D+Rectangle+%7B%0A++++++++++++width%3A+self.preferred-width%3B+%2F%2F+Should+actually+be+%60%40children%60%27s+size.%0A++++++++++++height%3A+self.preferred-height%3B%0A%0A++++++++++++%40children%0A++++++++%7D%0A++++%7D%0A%0A++++right+%3A%3D+Rectangle+%7B%0A++++++++x%3A+root.width+-+self.width%3B%0A++++++++y%3A+0%3B%0A++++++++width%3A+left.width%3B%0A++++++++background%3A+blue%3B%0A%0A++++++++right-inner+%3A%3D+Rectangle+%7B%0A++++++++++++width%3A+50px%3B%0A++++++++++++height%3A+50px%3B%0A++++++++++++background%3A+%23fff8%3B%0A++++++++%7D%0A++++%7D%0A%7D%0A%0Aexport+component+Demo+%7B%0A++++Triad+%7B%0A++++++++width%3A+100%25%3B%0A%0A++++++++Button+%7B%0A++++++++++++text%3A+%22Foo%22%3B%0A++++++++%7D%0A++++%7D%0A%7D%0A&style=fluent) shows the use case that made me desire the feature. Change the size of the preview to see what it's about. In my app, there are control buttons that must always be centered, no matter what's displayed to their left or right. To the left of the buttons, there *may* be informational `Text`s, right-aligned in their block (so, directly next to the buttons). The space to the right of the buttons is used to show error messages, if needed.

**Comment by ogoffart** [comment]
There is many usecases:
 - something like a SplitterView of some sort (QSplitter) that can be used to resize elements inside it
 - generalize the TabWidget that currently rely on compiler support.
 - Having customization point for theming
 - ...

We need to think of a syntax for different things.

 - Loop through children.  (eg `for X in @children : SomeComponent { X{} }`
 - Maybe `@children[0]` for the first children. How to make children mandatory or optional?
 - Named children: `@children(foobar)` and then `Foo { foobar |= Rectangle { ... }   }`  (not using quotes here because it should be identifiers)
 - We need to be able to constraint a children to an interface (#1870) 

(All my proposed syntax there are just some quick example, but nicer syntax need to be sorted out.)

**Comment by Enyium** [comment]
> We need to think of a syntax for different things.
> 
> * Loop through children.  (eg `for X in @children : SomeComponent { X{} }`
> * Maybe `@children[0]` for the first children. How to make children mandatory or optional?

This actually involves *batches* of children. The current `@children` would directly map to the new `@children["default"]`/`@children(default)`. Each slot could be filled with multiple siblings of children, as you see in my code above, just like this is now the case with the one and only slot possible `@children`. You should also be able to specify multiple non-`Slot` elements with `slot` properties in succession using the same name, which would be identical to provide a `Slot` with multiple children.

I'm not sure how iterating over children would fit into this proposal. Different slot contents should be able to be placed anywhere in the internal tree of a component, even in a different order than at the usage site. If it would ever be useful to use a `for` loop with children, I think this would be separate issue.

I admit index access like `@children[1]` meaning a batch of children could be confusing. Maybe named slots would be enough.

---

With Vue.js, you can even pass slot-specific data out to the slot usage site. See [here](https://vuejs.org/guide/components/slots.html#scoped-slots). I currently don't know what could possibly be the cases where one would need this with Slint, however.

If you'd want to implement this also, at the component usage site, this could look like:

```slint
Slot {
    name: "footer";
    // Or simply `in property` without `slot`.
    slot in property <int> name-from-component-but-unsuitable-because-already-in-use as received;
    Text { text: "The Footer\n\{received}"; }
}
```

In the component, it could perhaps look like the following:

```slint
@children(footer) {
    out property <int> the-name: 123;
}
```

Maybe, `out` in the component and `in` at the usage site would be the only variation that'd make sense, so everything else should be disallowed?

**Comment by SK83RJOSH** [comment]
I personally have to wonder if children shouldn't be properties to begin with – then you could explicitly communicate if it supports multiple, one, or if they're optional. Or even override default children implementations. It would be largely the same as assigning named elements is now as well.

One further benefit would be that accessing them would be available outside of slint, using the same syntax that other properties already use. Meaning it could be a bit easier to do dynamic views, like a stack based UI for example, without needing models.

Though perhaps you have good reasons for disallowing all of this – it just seems to me like a nice fit. :slightly_smiling_face:

**Comment by Enyium** [comment]
> accessing them would be available outside of slint, using the same syntax that other properties already use.

This reminds me of #2390. But I also read something about them not wanting to provide access to children, because not doing so gives them more freedom to optimize the Slint code.

**Comment by SK83RJOSH** [comment]
Aye, I imagine there's good arguments against doing it, and I was unaware of that particular issue - but indeed that's precisely what I was talking about. It's a use case we're going to have eventually, as we're planning to use interpreted slint in our project and let users manipulate the UI in a plug-n-play sort of fashion (hopefully). 🙂

**Comment by tronical** [comment]
If the children were exposed as a model or array at runtime, that would be cool. But what should the ownership be? Is a property a strong reference? (Risk of leaks through circular references) Is it a weak reference? (What happens if the object is deleted during binding evaluation)

I think if we can find good answers then that might level us up :)

**Comment by SK83RJOSH** [comment]
I would see things staying mostly as they are now, and user-defined properties can't be `element ref`. This nips any issues with circular references in the bud, I think?

Though you bring up very good points, and I'm left wondering, how does `focus-scope` avoid these issues right now? Is it simply impossible to create circular references due to the current constraints?

My proposal would be that users can only define properties which *own* the underlying element, and are neither strong nor weak references. These owning properties will always be `Copy`/`Clone`/`Default` constructed. Though they very well could be strong references under the hood, but non-internal code would be disallowed from constructing them as such.

Thus allowing the following:
```slint
component Page {
    // in lieu of optional, these would be default constructed
    in-out property <PageHeader> header;
    in-out property <PageBody> body;
    in-out property <[PageButton]> buttons;

    // using the same syntax for expansion
    VerticalLayout {
        @header
        @body
    }
    HorizontalLayout {
        // could be simplified as `@buttons`
        for button in buttons {
           @button
        }
    }
}

component HomePage inherits Page {
    header := PageHeader {
        text: "Hello, World!";
    }
    body := PageBody {
        text: "Lorem ipsum dolor sit amet"
    }
    buttons := [
        PageButton { text: "1"; },
        PageButton { text: "2"; },
        PageButton { text: "3"; },
    ]
}
```

I would then propose, that `children` becomes an explicit property on `SlintElement`:
```slint
struct SlintElement {
    in-out property <[SlintElement]> children;
}
```

The current "magic" of coalescing all children into that array is kept, as is the current `@` syntax for expansion. It might be nice to also support a less verbose syntax for assigning arrays of elements, though I don't see that as being strictly required.

Though these are just my off the cuff thoughts, and perhaps there are still issues left unresolved? My apologies if this is derailing this thread by the way - happy to move this discussion elsewhere if you'd like. 🙂

**Comment by NigelBreslaw** [comment]
I've seen this offered elsewhere with the following syntax:

```slint
// MyMagicButton.zing
Item {
    alias icon: innerIcon;
    alias moreIcons: moreInnerIcons;

    innerIcon := Item {

    }

    moreInnerIcons := Column {

    }
}

// Using the component
MyMagicButton {
    icons {
        Image { source: "../myicon.png"};
    } 
    moreIcons {
        Image { source: "../myicon.png"};
    }
}
```

Where I think `alias` is conceptually similar to `<=>` but at least for me clearer. "Expose this internal property/callback directly, but with a new name". 

This would make adding say an icon to a button easy. However in Zing this concept is aided by visual items having a 'children' property. This is a read only array of children the item has. Something that has many uses! But in this case it can be used to allow items to have a default implementation that is not used if the component has some children added. 

```
export component ExportedComponent {
    alias icons: headerIcons;

    headerIcons := VerticalLayout {}

    if headerIcons.children.length == 0 : Rectangle {
        ... default icon or whatever
    }
}
```

I saw talk about being able to type or enforce interfaces which sounds next level amazing. To think of a future where we have some kind of Navigation components that will allow nice transitions between different views. But able to enforce each component someone adds actually is a View or whatever sounds so useful.

**Comment by TheColorRed** [comment]
As someone who comes from `Angular`, I would like to chime in on another type of syntax that is partially similar to angular.

Here we create three components for the example: `Header`, `Footer`, and `Paragraph`.

````slint
export component Header {
  in property: text;
  Text { text: text; }
}
export component Footer {
  in property: text;
  Text { text: text; }
}
export component Paragraph {
  in property: text;
  Text { text: text; }
}
````

We then use another component that renders the three items where children (when used with a parameter) would receive a component type as an argument (or even a list of components). When no arguments are passed, all the remaining components that didn't match a selector, would render to that slot.

1. First select all the components and put them into their respective group as an array (in this case a group to hold header components and a group to hold footer components).
2. Next place all the remaining components into an miscellaneous group as an array.
3. Render them in the order that they were defined in the layout into the slot.

```slint
export component Main {
    HorizontalLayout {
        // Can only have one
        @children(Header);
    }
    VerticalLayout {
        // Can only have one
        @children
    }
    HorizontalLayout {
        // Can only have one
        @children(Footer);
    }
}
```

Now the last thing to do is create the main component with its content maybe something like this:

```slint
export component App inherits Window {
  Main {
    Header { text: "Welcome"; }
    Footer { text: "Goodbye"; }
    Paragraph { text: "Lorem ipsum odor amet, consectetuer adipiscing elit." }
    Paragraph { text: "Lorem ipsum dolor sit amet." }
    Button { text: "Hello World"; }
    Footer { text: "See you next time!"; }
  }
}
```

The rendered output would look like this. Note that the two footers were originally out of order, and now they are grouped together:

```slint
export component Main {
    HorizontalLayout {
      Header { text: "Welcome"; }
    }
    VerticalLayout {
      Paragraph { text: "Lorem ipsum odor amet, consectetuer adipiscing elit." }
      Paragraph { text: "Lorem ipsum dolor sit amet." }
      Button { text: "Hello World"; }
    }
    HorizontalLayout {
      Footer { text: "Goodbye"; }
      Footer { text: "See you next time!"; }
    }
}
```

Now angular also has reusable templates that you can reuse since you can on use `@children(xxx)` one time, however the template can be reused:

```slint
export component Main {
    HorizontalLayout {
      @template(MyTemplate);
    }
    VerticalLayout {
      @template(MyTemplate);
    }
    HorizontalLayout {
      @template(MyTemplate);
    }
}

template MyTemplate {
    HorizontalLayout {
      @children(MyCompnent);
    }
}
```

**Comment by tronical** [comment]
My preferred syntax for this this feature  of declaring an interface with certain constraints:

```slint

component Header { in property <string> title; ... } // could also be a sub-component of Page
component Footer { in property <string> footnotes: ... } 

component Page {
    // Declare the interface
    @children {
        header: Header,
        footer: Footer,
        ...
    }

    
    // Implementation
    VerticalBox {
        SometingThatWrapsAroundTheHeader { @children.header }
        ScrollView { @children } // the body
        @children.footer
    }
}

// Use-site:

    Page {
        header: Header { ... }
        footer: Footer { ... }
        TextEdit { ... }
    }
```

**Comment by ogoffart** [comment]
Thanks @ThePartyPinguin for making a PR (#10738) prototype, which makes us continue the discussion we have on the design of this feature

We also currently have an experimental interface feature in the Slint language which we could also make of.

We need:
 1. a syntax to declare slots in a component (declare the interface of that component)
 2. a syntax to use these slots in that component.
 3. a syntax to instantiate these when using that components in the use-site

(1 and 2 could be merged if we say that just using slots declare their use)

Depending on how we do that, we could end up deprecating `@children` if this solution is strictly more generic.

Iterating over the previous comment from @tronical 

```slint
// Interface is an experimental feature already in Slin's master branch.
interface Header {
    in property <string> title;
    callback clicked();

    // Extension? Should interface have slots themselves.
    slot something;
}

component Page {
    // Declare the interface of Page
    slot <Header> header; // <- `Header` is an interface
    slot footer; // no interface specified
    default slot children;

    VerticalBox {
        // Use a slot as if it was a component
        SometingThatWrapsAroundTheHeader { header {} }
        ScrollView { children {} } // the body
        footer {}
    }
}

component UseSite {
    Page {
        // Use `*<slotname>:` to put a named children
        *header: my-header := Header { title: "Hello"; }

        // There can be several footers
        *footer: SomeElement {}
        *footer: OtherElement {}

        TextEdit {}
    }
}

```


How to implement the TabWidget (need a few more features)

```slint
export interface Tab {
    in property <string> title;
}
component TabWidget {
    // feature: default slot is the @children
    default slot <Tab> tabs;
    in-out property <int> current-index;

    VerticalLayout {
        TabBarInternal {
            // feature: `.count` for repeated element
            for idx in tab_inst.count: TabInternal {
                // feature: access a specific child in a repeater
                title: tab_inst[idx].title;
                current: idx == current-index;
                clicked => {  current-index = idx; }
            }

            /* Alternative: access slots directly
            for idx in tabs.count: TabInternal {
                // feature: access a specific child in a repeater
                title: tabs[idx].title;
            }
            */
        }
        Rectangle {
            border-color: Palette.border;
            border-width: 1px;
            // Feature `for` loop over a slot
            for x[idx] in tabs: tab_inst := Rectangle {
                property <string> title: tab.title;
                visible: idx == current-index;
                // feature: instantiate from a variable
                tab := x {}
            }
        }
    }
}
export component Tab implements Tab {
    @children
}
```


A Splitter
```slint
export component Splitter {
   default slot children;

    l := HorizontalLayout {
        for x[idx] in children: r := Rectangle {
            if idx < children.count: SplitterSeparator {
                x: r.width - self.width / 2;
                moved(deltapx) => {
                    // this untested formula is the complex part of the splitter but out of the scope of this feature
                    let u = (l.width * r.horizontal-stretch / l.horizontal-stretch + deltapx) / l.width;
                    r.horizontal-stretch = u * (l.horizontal-stretch - r.horizontal-stretch) / (1 - u);
                }
            }
            x{ }
        }
    }
}

```

All of this is just rough suggestions we will probably iterate over again.

**Comment by LeonMatthes** [comment]
A note for further discussion. This syntax:
`default slot <Tab> tabs;` is very close to `default property <Tab> tabs`.

Is a slot so conceptually different to a property that it warrants introducing a new syntax?
Could these two concepts just be one and the same?

They might well be different, but we need to be clear on the reasonong here.

**Comment by ThePartyPinguin** [comment]
This syntax makes a lot of sense to me:
```
// Interface is an experimental feature already in Slin's master branch.
interface Header {
    in property <string> title;
    callback clicked();

    // Extension? Should interface have slots themselves. <-- Probably yes if we view slots as properties of an element
    slot something;
}

component Page {
    // Declare the interface of Page
    slot <Header> header; // <- `Header` is an interface
    slot <Rectangle> sub_header; // <-- Any 'Rectangle' or 'inheritor' is allowed to be the sub_header;
    slot footer; // no interface specified
    default slot children;

    VerticalBox {
        // Use a slot as if it was a component
        SometingThatWrapsAroundTheHeader { header { // Can we also update properties of header here? Should we allow it? } }
        ScrollView { children {} } // the body
        footer {}
    }
}

component UseSite {
    Page {
        // Use `*<slotname>:` to put a named children
        *header: my-header := Header { title: "Hello"; }

        // There can be several footers
        *footer: SomeElement {}
        *footer: OtherElement {}

        TextEdit {}
    }
}
```
Incorporating the interface feature (even though I haven't looked at it in detail), is the way to go here I think yet.
But allowing a component that `inherits` (from a `Rectangle` in case of `sub_header`) should also be possible then?

Also that a slot can have multiple items would be nice, but the syntax for that looks a bit confusing to me.
In you second example with the repeated element. Does this imply that a slot can always have multiple elements?
```
// feature: `.count` for repeated element
for idx in tab_inst.count: TabInternal { <-- not sure where `tab_inst` is defined in you example, is it implicitly generates?
    // feature: access a specific child in a repeater
    title: tab_inst[idx].title;
    current: idx == current-index;
    clicked => {  current-index = idx; }
}
```
This would be fine if we don't define an interface for a slot and just declare it as an _anonymous_ `slot`, with such a `slot` we also don't know what the contents really are so being able iterate over them or index them feels a bit useless(?). But when we declare it _with_ an interface, I think we should make it possible for the user to define it as either `slot<T>` for a single element or `slot<[T]>` for multiple elements. Which would be in line with how we define an array in a property `property<[T]>`

In this case `slot<[T]>` does have a `.count` and can be index `[i]` but `slot<T>` does not.

I would also like to try and do away with this 'double' looking assignment:
```
*header: my-header := Header { title: "Hello"; }
         ^^^^^^^^^
```
and just use:
```
property<length>: *header.width;

*header:= Header { title: "Hello"; }
```

For multiple slots we could use;
```
property<length> first_footer_height: *footer[0].height;
property<length> second_footer_height: *footer[1].height;

*header[0]:= SomeElement { }
*header[1]:= OtherElement { }
```
Although still not really sure how I feel about this one..

The `default slot` would be a stated before an _anonymous_ slot. Allowing for iteration and indexing.

---
I would also like to touch on the `for` syntax, with this example;
```
export component Splitter {
   default slot children;

    l := HorizontalLayout {
        for (x, r)[idx] in children:= Rectangle { <-- move the element to a tuple like (value, element) declaration
            if idx < children.count: SplitterSeparator {
                x: r.width - self.width / 2;
                moved(deltapx) => {
                    // this untested formula is the complex part of the splitter but out of the scope of this feature
                    let u = (l.width * r.horizontal-stretch / l.horizontal-stretch + deltapx) / l.width;
                    r.horizontal-stretch = u * (l.horizontal-stretch - r.horizontal-stretch) / (1 - u);
                }
            }
            x{ }
        }
    }
}
```

**Comment by ThePartyPinguin** [comment]
@LeonMatthes 
> A note for further discussion. This syntax: `default slot <Tab> tabs;` is very close to `default property <Tab> tabs`.
> 
> Is a slot so conceptually different to a property that it warrants introducing a new syntax? Could these two concepts just be one and the same?
> 
> They might well be different, but we need to be clear on the reasonong here.

You're saying we just allow `in property<ElementType|Interface> prop_name`?

**Comment by ogoffart** [comment]
@LeonMatthes, @ThePartyPinguin @tronical  and I had a discussion where we brainstormed possible syntax.

```slint
component Page {
    // 1 : declaration

    slot <Header> header; // <- `Header` is an interface <- allows for a single assignment
    // potential "optional syntax", not a near-term target  
    slot <[Header]> headers; // <- `Header` is an interface <- allows for a multiple assignment

    slot header; // without specifying an interface, any single component goes    
    slot <component> header; // same as before, component is a reserved interface name
    slot <[component]> headers; // multiple components

    default slot <[component]> children; // implicitly declared when using @children

    // 2: instantiate

    // instantiate the single component
    header {}
    // It is mandatory to instantiate the component once, and only once.
    //  (and not in `for` or `if`)

    // for multiple components:
    for header in headers: FooBar { header{} }

    // @chilren is equivalent to 
    for c in children: c {}
}

// 3: Use (this is the most controversial right now)
component UseSite {
    Page {
        header << Foo { 
             // ...
        }

        headers << [
            foo1 := Foo { 
                 //...
            }
            Foo {
               // ...
            }
        ]
    }
}
```


So use `slot` to declare slot, and `<<` operator at the use.

<details>
<summary>Here is some random stuff from our notes</summary>

```slint
component UseSite {

    function hello() {
        page1.*hello = @element(
            Header {
                title: "Hello";
            }
        );
    }

    page1 := Page {
        // Note: We may need to have a special character to indicate slots
        // for our parser.
        // From a user perspective we may not want them.
        // TODO: May or may not need semicolons.

        // ### CATEGORY 1: property-like assignment
        // Easy to use, may be confusing, hard to disambiguate in the parser
        header: Foo {};

        // Use `*<slotname>:` to put a named children (e.g. some kind of special character)
        *header: [
            if foo : Foo {}
            header := Header { title: "Hello"; }
        ];
        *header: Foo {};
        // maybe this is best if we want to have a special character
        // makes the @ even more ambiguous! :(
        @header: Foo {};

        // Or it could be some kind of special assignment operator
        // or maybe <=
        header << Foo {
            // ..
        }
        header << [
            Foo {
            // ..
            }
            Foo {
            // ..
            }
        ]

        for i in [1,2 3,4,5,6]: [
            Rectangle {}
            Rectangle {}
        ]
        // This is already used, so maybe not a good idea
        header := Foo {};

        // Use the @ operator for "custom literals"
        header: @elem([
            if foo: Foo {}
        ]);

        // Group slots
        slots [
            header: Foo {};
        ]

        states [

        ]


        // ### CATEGORY 2: Assign at use site?
        Foo(slot:header) {}

        my-header := Foo {}
        slots {
            // Problematic: Moves the element, currently this never happens
            header: my-header;
        }

        header <= Foo


        // There can be several footers
        // We probably don't want this!
        *footer: SomeElement {}
        *footer: OtherElement {}

        foo: foo := foo {}
        foo := foo: foo {}
        foo := *header
        *header: [

        ]
        
}
```

</details>

We also discussed that the `[ elements ]` syntax could be used for `for` and `if` yielding several items. But that's another unrelated issue.

**Comment by LeonMatthes** [comment]
>     // It is mandatory to instantiate the component once, and only once.
>     //  (and not in `for` or `if`)

I feel like it should work in an `if`, but definitely not in a `for`.

**Comment by ogoffart** [comment]
> I feel like it should work in an if, but definitely not in a for.

We need to reject it everywhere `@children` is rejected. That is in `for`,  `if`, `PopupWindow`, ...
That's because right now we would allow to access items within component that might not be instantiated.  We may want to relax some rules but it needs to be done carefully.

**Comment by eira-fransham** [comment]
Could we make the restriction to one or many components be transparent at use-site, to make it easier to be forwards-compatible and simplify the syntax?

```slint
component Foo {
  // Maybe `slot my-slot` could be shorthand for `slot <component> my-slot`
  slot <component> my-slot;
  slot <[component]> many-slot;

  my-slot {}
  for elem in many-slot: elem {}
}

component Bar {
  Foo {
    my-slot {
      Text { text: "Hello, world!" }
    }

    many-slot {
      Text { text: "Hello," }
      Text { text: "world!" }
    }
  }
}
```

Personally, I quite like the way it makes the use of default and non-default slots as similar as possible. I don't think it makes the parsing context-sensitive, but it does mean that we need to change the semantics based on the result of name lookup.

I will say though, whatever syntax we use, I think that both the `<<` syntax and the use of `[ .. ]` with components is very unintuitive.

**Comment by LeonMatthes** [comment]
Hm, it's true that this doesn't make the parsing itself context-sensitive, as it just parses as an element with children.

And I agree that this looks more familiar from a glance.
What I would be worried about is that this looks too much like a normal element instantiation, so might be confusing for that reason.

I suggest that we merge the initial PR with the `<<` and `[]` syntax, and then we open a separate PR that refactors it to this proposed syntax, to see what the examples would look like then, and how much this complicates the compiler.
Definitely willing to try this, but I wouldn't want to hold up the existing Slots PR for it atm. We want to figure out the semantics of it first.

**Comment by eira-fransham** [comment]
@LeonMatthes Yeah, looking like element instantiation is my biggest concern. In context, the different naming convention is probably enough that it's not a big problem, but it's a shame that that's not enforced by the syntax and it could be confusing when you first see it. I feel pretty strongly against the existing syntax, but if it's already implemented then maybe better to just get it merged for now and revisit it later, as you say.
